src/servers/src/postgres/handler.rs

```diff
@@ -177,10 +245,105 @@ fn type_translate(origin: &ConcreteDataType) -> Result<Type> {
     }
 }
 
+fn type_pg_to_gt(origin: &Type) -> Result<ConcreteDataType> {
+    // Note that we only support a small amount of pg data types
+    match origin {
+        &Type::BOOL => Ok(ConcreteDataType::boolean_datatype()),
+        &Type::CHAR => Ok(ConcreteDataType::int8_datatype()),
+        &Type::INT2 => Ok(ConcreteDataType::int16_datatype()),
+        &Type::INT4 => Ok(ConcreteDataType::int32_datatype()),
+        &Type::INT8 => Ok(ConcreteDataType::int64_datatype()),
+        &Type::VARCHAR | &Type::TEXT => Ok(ConcreteDataType::string_datatype()),
+        &Type::TIMESTAMP => Ok(ConcreteDataType::timestamp_datatype(
+            common_time::timestamp::TimeUnit::Millisecond,
+        )),
+        &Type::DATE => Ok(ConcreteDataType::date_datatype()),
+        &Type::TIME => Ok(ConcreteDataType::datetime_datatype()),
+        _ => error::InternalSnafu {
+            err_msg: format!("unimplemented datatype {origin:?}"),
+        }
+        .fail(),
+    }
+}
+
+#[derive(Default)]
+pub struct POCQueryParser;
+
+impl QueryParser for POCQueryParser {
+    type Statement = (Statement, String);
+
+    fn parse_sql(&self, sql: &str, types: &[Type]) -> PgWireResult<Self::Statement> {
+        let mut stmts = ParserContext::create_with_dialect(sql, &GenericDialect {})
+            .map_err(|e| PgWireError::ApiError(Box::new(e)))?;
+        if stmts.len() != 1 {
+            Err(PgWireError::UserError(Box::new(ErrorInfo::new(
+                "ERROR".to_owned(),
+                "42P14".to_owned(),
+                "invalid_prepared_statement_definition".to_owned(),
+            ))))
+        } else {
+            let mut stmt = stmts.remove(0);
+            if let Statement::Query(qs) = &mut stmt {
+                for t in types {
+                    let gt_type =
+                        type_pg_to_gt(t).map_err(|e| PgWireError::ApiError(Box::new(e)))?;
+                    qs.param_types_mut().push(gt_type);
+                }
+            }
+
+            Ok((stmt, sql.to_owned()))
+        }
+    }
+}
+
+fn parameter_to_string(portal: &Portal<(Statement, String)>, idx: usize) -> PgWireResult<String> {
+    let param_type = portal.statement().parameter_types().get(idx).unwrap();
```

Comment:
Is it safe to unwrap here?

Reply:
Yes, the `idx` is from our range loop. I've added a comment for this.